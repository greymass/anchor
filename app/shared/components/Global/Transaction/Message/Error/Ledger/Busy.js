import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Header, Icon, List, Segment } from 'semantic-ui-react';
import ReactJson from 'react-json-view';

export class GlobalTransactionMessageErrorLedgerBusy extends Component<Props> {
  render() {
    const {
      error,
      t
    } = this.props;
    return (
      <Segment basic color="red" size="large">
        <Header size="medium">
          <Icon name="warning sign" />
          <Header.Content>
            Ledger Device Busy
            <Header.Subheader>
              The Ledger device is currently trying to process a different transaction.
            </Header.Subheader>
          </Header.Content>
        </Header>
        <p>Before trying this transaction again, address this issue by taking one or more of the following steps:</p>
        <List relaxed bulleted>
          <List.Item>Clear any pending transactions on the Ledger device.</List.Item>
          <List.Item>Reconnect the Ledger to the computer.</List.Item>
        </List>
        {(typeof error === 'object' && error.json) ? (
          <ReactJson
            collapsed={1}
            displayDataTypes={false}
            displayObjectSize={false}
            iconStyle="square"
            name={null}
            src={error.json.error}
            style={{ padding: '1em' }}
          />
        ) : ''}
      </Segment>
    );
  }
}

export default translate('global')(GlobalTransactionMessageErrorLedgerBusy);
