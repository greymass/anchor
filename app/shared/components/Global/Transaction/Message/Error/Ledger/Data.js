import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Header, Icon, List, Segment } from 'semantic-ui-react';
import ReactJson from 'react-json-view';

export class GlobalTransactionMessageErrorLedgerData extends Component<Props> {
  render() {
    const {
      error,
      t
    } = this.props;
    return (
      <Segment color="red" size="large">
        <Header size="medium">
          <Icon name="warning sign" />
          <Header.Content>
            Enable "Contract Data" on Ledger device
            <Header.Subheader>
              In order to sign this transaction, a setting needs to be enabled on the Ledger device.
            </Header.Subheader>
          </Header.Content>
        </Header>
        <p>To enable this setting:</p>
        <List relaxed bulleted>
          <List.Item>Open the EOS app on your Ledger device.</List.Item>
          <List.Item>Cycle through the menu and find the "Settings" option.</List.Item>
          <List.Item>Press both buttons to enter the settings list.</List.Item>
          <List.Item>In settings, cycle through the items and find the "Contract Data" option.</List.Item>
          <List.Item>Press both buttons to change the option until it says "Allowed" at the bottom.</List.Item>
          <List.Item>Cycle again through the settings to find the "Back" button, then click both buttons to return to the main interface.</List.Item>
          <List.Item>With this new setting enabled, try your transaction again.</List.Item>
        </List>
        {(typeof error === 'object' && error.json) ? (
          <ReactJson
            collapsed={1}
            displayDataTypes={false}
            displayObjectSize={false}
            iconStyle="square"
            name={null}
            src={error.json.error}
            style={{ padding: '1em' }}
          />
        ) : ''}
      </Segment>
    );
  }
}

export default withTranslation('global')(GlobalTransactionMessageErrorLedgerData);
