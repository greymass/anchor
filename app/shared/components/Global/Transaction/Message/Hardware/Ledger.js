// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Header, Icon, Segment } from 'semantic-ui-react';

export class GlobalTransactionMessageHardwareLedger extends Component<Props> {
  render() {
    const {
      t
    } = this.props;
    return (
      <Segment basic>
        <Header icon size="large" textAlign="center">
          <Icon
            loading
            name="circle notched"
          />
          {t('global_transaction_hardware_ledger_awaiting_device_header')}
          <Header.Subheader
            content={t('global_transaction_hardware_ledger_awaiting_device_subheader')}
          />
        </Header>
      </Segment>
    );
  }
}

export default translate('global')(GlobalTransactionMessageHardwareLedger);
