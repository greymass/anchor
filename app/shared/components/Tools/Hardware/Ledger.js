// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import { Button, Icon, Header, Label, List, Segment, Table } from 'semantic-ui-react';

import ToolsHardwareLedgerStatus from './Ledger/Status';
import HardwareLedger from '../../../utils/Hardware/Ledger';

class ToolsHardwareLedger extends Component<Props> {
  displayPublicKey = () => {
    this.props.actions.ledgerGetPublicKey(0, true);
  };
  getPublicKey = () => {
    this.props.actions.ledgerGetPublicKey(0);
  };
  render() {
    const {
      accounts,
      actions,
      ledger,
      settings,
      status,
      t,
    } = this.props;
    const account = accounts[settings.account];
    const { transport } = new HardwareLedger();
    if (!account) return false;
    return (
      <Segment color="violet" piled style={{ margin: 0 }}>
        <ToolsHardwareLedgerStatus
          actions={actions}
          ledger={ledger}
          settings={settings}
          status={status}
        />
        {(status !== 'connected')
          ? (
            <Segment padded size="large">
              <Header>
                {t('ledger_unavailable_text_header')}
              </Header>
              <p>{t('ledger_unavailable_text_1')}</p>
              <List>
                <List.Item>
                  <Icon name="right triangle" />
                  <List.Content>
                    <List.Description>
                      {t('ledger_unavailable_list_1')}
                    </List.Description>
                  </List.Content>
                </List.Item>
                <List.Item>
                  <Icon name="right triangle" />
                  <List.Content>
                    <List.Description>
                      {t('ledger_unavailable_list_2')}
                    </List.Description>
                  </List.Content>
                </List.Item>
                <List.Item>
                  <Icon name="right triangle" />
                  <List.Content>
                    <List.Description>
                      {t('ledger_unavailable_list_3')}
                    </List.Description>
                  </List.Content>
                </List.Item>
                <List.Item>
                  <Icon name="right triangle" />
                  <List.Content>
                    <List.Description>
                      {t('ledger_unavailable_list_4')}
                    </List.Description>
                  </List.Content>
                </List.Item>
                <List.Item>
                  <Icon name="right triangle" />
                  <List.Content>
                    <List.Description>
                      {t('ledger_unavailable_list_5')}
                    </List.Description>
                  </List.Content>
                </List.Item>
              </List>
            </Segment>
          )
          : (
            <Segment padded size="large">
              <Header>
                {t('ledger_available_text_header')}
              </Header>
              <p>{t('ledger_available_text_1')}</p>
              <p>{t('ledger_available_text_2')}</p>
            </Segment>
          )
        }
        <Table definition>
          <Table.Header>
            <Table.HeaderCell colSpan="16">
              {t('ledger_status_table')}
            </Table.HeaderCell>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell width={6}>
                {t('ledger_row_automatic_detection')}
              </Table.Cell>
              <Table.Cell>
                {(settings.hardwareLedgerSupport)
                  ? <Label basic color="green" content="enabled" horizontal />
                  : <Label basic color="yellow" content="disabled" horizontal />
                }
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>
                {t('ledger_row_waiting_device')}
              </Table.Cell>
              <Table.Cell>
                {(ledger.listening)
                  ? <Label basic color="green" content="waiting" horizontal />
                  : <Label basic color="yellow" content="disabled" horizontal />
                }
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>
                {t('ledger_row_connected')}
              </Table.Cell>
              <Table.Cell>
                {(transport)
                  ? <Label basic color="green" content={t('ledger_row_connected_true')} horizontal />
                  : <Label basic color="yellow" content={t('ledger_row_connected_false')} horizontal />
                }
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>
                {t('ledger_row_software_version')}
              </Table.Cell>
              <Table.Cell>
                {(transport)
                  ? false
                  : <Label basic color="yellow" content={t('ledger_row_connected_false')} horizontal />
                }
                {(transport && ledger.application && ledger.application.version)
                  ? ledger.application.version
                  : false
                }
                {(transport && (!ledger.application || !ledger.application.version))
                  ? <Label basic color="orange" content={t('ledger_status_awaiting_application_subheader')} horizontal />
                  : false
                }
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>
                {t('ledger_row_public_key')}
              </Table.Cell>
              <Table.Cell>
                {(transport)
                  ? false
                  : <Label basic color="yellow" content={t('ledger_row_connected_false')} horizontal />
                }
                {(transport && ledger.publicKey)
                  ? (
                    <React.Fragment>
                      <p>{ledger.publicKey.wif}</p>
                      <Button
                        content={t('ledger_confirm_public_key')}
                        onClick={this.displayPublicKey}
                        primary
                      />
                    </React.Fragment>
                  )
                  : false
                }
                {(transport && !ledger.publicKey)
                  ? (
                    <React.Fragment>
                      <Button
                        content={t('ledger_retrieve_public_key')}
                        onClick={this.getPublicKey}
                        primary
                      />
                    </React.Fragment>
                  )
                  : false
                }
              </Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </Segment>
    );
  }
}

export default translate('ledger')(ToolsHardwareLedger);
