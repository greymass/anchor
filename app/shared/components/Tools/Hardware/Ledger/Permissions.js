// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import ReactJson from 'react-json-view';

import ToolsModalPermissionAuthSet from '../../Modal/Permissions/Auth/Set';
import WalletPanelLocked from '../../../Wallet/Panel/Locked';

import EOSAccount from '../../../../utils/EOS/Account';

import { Button, Checkbox, Container, Icon, Header, Label, List, Message, Popup, Segment, Table } from 'semantic-ui-react';

class ToolsHardwareLedgerPermissions extends Component<Props> {
  render() {
    const {
      account,
      actions,
      blockExplorers,
      keys,
      ledger,
      settings,
      system,
      t,
      validate,
      wallet,
    } = this.props;
    if (!account) return false;
    const { pubkey } = keys;
    const authorization = new EOSAccount(account).getAuthorization(pubkey);
    // Don't display permissions modification until ledger connection established
    if (!(ledger.devicePath && ledger.application)) return false;
    // Don't display while wallet is locked
    if (settings.walletMode !== 'watch' && !(keys && keys.key)) {
      return (
        <WalletPanelLocked
          actions={actions}
          settings={settings}
          validate={validate}
          wallet={wallet}
        />
      );
    }
    return (
      <Table definition>
        <Table.Body>
          <Table.Row>
            <Table.Cell>
              Account
            </Table.Cell>
            <Table.Cell>
              {settings.account}
            </Table.Cell>
          </Table.Row>
          {(account.permissions.map((data) => (
            <React.Fragment>
              {data.required_auth.keys.map((permission) => (
                <Table.Row key={`${data.perm_name}-${permission.key}`}>
                  <Table.Cell collapsing textAlign="right">{data.perm_name}</Table.Cell>
                  <Table.Cell>
                    {(ledger.publicKey && permission.key === ledger.publicKey.wif)
                      ? "Uses Ledger"
                      : "No"
                    }
                  </Table.Cell>
                  <Table.Cell verticalAlign="middle">
                    {permission.key}
                  </Table.Cell>
                  <Table.Cell>
                    {(!ledger.publicKey || permission.key === ledger.publicKey.wif)
                      ? false
                      : (
                        <React.Fragment>
                          {(
                            !authorization
                            || (data.perm_name === 'owner' && authorization.permission === 'owner')
                            || (data.perm_name !== 'owner')
                          )
                            ? (
                              <ToolsModalPermissionAuthSet
                                actions={actions}
                                auth={data}
                                blockExplorers={blockExplorers}
                                button={{
                                  color: 'purple',
                                  content: t('tools_modal_permissions_auth_use_ledger_button'),
                                  fluid: false,
                                  icon: 'lock',
                                  size: 'small'
                                }}
                                onClose={this.onClose}
                                newkey={ledger.publicKey.wif}
                                pubkey={pubkey}
                                settings={settings}
                                system={system}
                              />
                            )
                            : (
                              <Popup
                                content={t('tools_modal_permissions_auth_edit_button_disabled')}
                                inverted
                                position="top center"
                                trigger={(
                                  <Button
                                    content={t('tools_modal_permissions_auth_edit_button')}
                                    floated="right"
                                    icon="pencil"
                                    size="small"
                                  />
                                )}
                              />
                            )
                          }
                        </React.Fragment>
                      )
                    }
                  </Table.Cell>
                </Table.Row>
              ))}
            </React.Fragment>
          )))}
        </Table.Body>
      </Table>
    );
  }
}

export default translate('ledger')(ToolsHardwareLedgerPermissions);
