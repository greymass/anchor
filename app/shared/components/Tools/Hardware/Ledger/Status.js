// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Checkbox, Header, Icon, Segment } from 'semantic-ui-react';

class ToolsHardwareLedgerStatus extends Component<Props> {
  start = () => {
    this.props.actions.ledgerStartListen();
  }
  stop = () => {
    this.props.actions.ledgerStopListen();
  }
  restart = () => {
    this.props.actions.ledgerStopListen();
    setTimeout(() => {
      this.props.actions.ledgerStartListen();
    }, 500);
  }
  toggleDetection = (e, { checked }) => {
    this.props.actions.setSetting('hardwareLedgerSupport', checked);
  }
  render() {
    const {
      hideOptions,
      ledger,
      settings,
      status,
      t,
    } = this.props;

    return (
      <React.Fragment>
        {(ledger.listening)
          ? (
            <Segment attached="top" color={(status !== 'connected') ? 'orange' : 'green'}>
              <Header icon size="large" textAlign="center">
                {(status === 'transport_error')
                  ? (
                    <Icon
                      color="orange"
                      name="warning sign"
                    />
                  )
                  : false
                }
                {(status !== 'transport_error' && status !== 'connected')
                  ? (
                    <Icon
                      loading
                      name="circle notched"
                    />
                  )
                  : false
                }
                {(status === 'connected')
                  ? (
                    (
                      <Icon
                        color="green"
                        name="usb"
                      />
                    )
                  )
                  : false
                }
                {t(`ledger_status_${status}_header`)}
                <Header.Subheader
                  content={t(`ledger_status_${status}_subheader`)}
                />
              </Header>
            </Segment>
          )
          : (
            <Segment attached="top" color="grey">
              <Header icon size="large" textAlign="center">
                <Icon
                  name="usb"
                />
                {t('ledger_status_disabled_header')}
                <Header.Subheader
                  content={t('ledger_status_disabled_subheader')}
                />
              </Header>
            </Segment>
          )
        }
        <Segment attached textAlign="center">
          {(status === 'transport_error')
            ? (
              <Button
                basic
                content={t('ledger_service_restart')}
                color="orange"
                icon="redo"
                onClick={this.restart}
              />
            )
            : false
          }
          {(status !== 'transport_error' && ledger.listening)
            ? (
              <Button
                basic
                content={t('ledger_service_stop')}
                color="orange"
                icon="stop"
                onClick={this.stop}
              />
            )
            : false
          }
          {(status !== 'transport_error' && !ledger.listening)
            ? (
              <Button
                color="green"
                content={t('ledger_service_start')}
                icon="play"
                onClick={this.start}
              />
            )
            : false
          }
        </Segment>
        {(status !== 'transport_error' && !hideOptions)
          ? (
            <Segment attached textAlign="center">
              <Checkbox
                defaultChecked={settings.hardwareLedgerSupport}
                label={t('ledger_service_automatic')}
                onChange={this.toggleDetection}
              />
            </Segment>
          )
          : false
        }
      </React.Fragment>
    );
  }
}

export default translate('ledger')(ToolsHardwareLedgerStatus);
