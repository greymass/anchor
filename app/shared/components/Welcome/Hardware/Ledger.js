// @flow
import React, { Component } from 'react';
import { Header, Segment } from 'semantic-ui-react';
import { translate } from 'react-i18next';
import WelcomeHardwareLedgerContainer from '../../../containers/Welcome/Hardware/Ledger';

class WelcomeHardwareLedger extends Component<Props> {
  render() {
    const {
      onClose,
      onComplete,
      onStageSelect,
      t
    } = this.props;
    return (
      <React.Fragment>
        <Header size="huge">
          <Header.Subheader>
            {t('welcome_step')} #2
          </Header.Subheader>
          {t('welcome_stage_account_ledger')}
        </Header>
        <Segment
          padded
          size="small"
          stacked
        >
          <p>{t('welcome_instructions_8')}</p>
          <WelcomeHardwareLedgerContainer
            onClose={onClose}
            onComplete={onComplete}
            onStageSelect={onStageSelect}
          />
        </Segment>
      </React.Fragment>
    );
  }
}

export default translate('welcome')(WelcomeHardwareLedger);
