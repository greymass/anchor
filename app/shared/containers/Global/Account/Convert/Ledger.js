// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { find } from 'lodash';
import { Button, Form, Header, Icon, Message, Modal, Segment, Table } from 'semantic-ui-react';
import TimeAgo from 'react-timeago';

import * as AccountActions from '../../../../actions/accounts';
import * as HardwareLedgerActions from '../../../../actions/hardware/ledger';
import * as SettingsActions from '../../../../actions/settings';
import * as SystemStateActions from '../../../../actions/system/systemstate';
import * as TransactionActions from '../../../../actions/transaction';
import * as UpdateAuthActions from '../../../../actions/system/updateauth';
import * as WalletActions from '../../../../actions/wallet';
import * as WalletsActions from '../../../../actions/wallets';

import WalletPanelButtonBroadcast from '../../../../components/Wallet/Panel/Button/Broadcast';
import ToolsModalPermissionAuthSet from '../../../../components/Tools/Modal/Permissions/Auth/Set';

import EOSAccount from '../../../../utils/EOS/Account';

class GlobalAccountConvertLedger extends Component<Props> {
  state = {
    keyIndex: 0,
  }
  componentDidMount() {
    const { status } = this.props;
    if (status === 'connected') {
      this.loadKey();
    }
  }
  completePrepare = () => {
    const {
      account,
      actions,
      authorization,
      wallets,
    } = this.props;
    const wallet = find(wallets, {
      account,
      authorization,
    });
    if (wallet && wallet.convertParameters) {
      const {
        path,
        publicKey
      } = wallet.convertParameters;
      const {
        wif
      } = publicKey;
      actions.completeConvertToLedger(
        account,
        authorization,
        wif,
        path
      );
      this.onClose();
    }
  }
  onChange = (e, { value }) => this.setState({ keyIndex: parseInt(value, 10) });
  onClose = () => {
    this.reset();
    this.props.onClose();
  }
  onFormSubmit = () => {
    const {
      account,
      actions,
      authorization,
      ledger
    } = this.props;
    const {
      path,
      publicKey
    } = ledger;
    this.setState({
      awaitingUpdate: true
    });
    actions.prepareConvertToLedger(account, authorization, path, publicKey);
  }
  loadKey = () => {
    const { actions } = this.props;
    actions.ledgerGetPublicKey(this.state.keyIndex);
  }
  reset = () => {
    const { actions } = this.props;
    actions.ledgerResetPublicKey();
    this.setState({
      keyIndex: 0
    });
  }
  stopPrepare = () => {
    const {
      actions,
      account,
      authorization
    } = this.props;
    actions.prepareConvertToLedgerAbort(account, authorization);
  }
  updateAccount = () => {
    const { account, actions } = this.props;
    actions.getAccount(account);
  }
  verify = () => {
    const { actions } = this.props;
    actions.ledgerGetPublicKey(this.state.keyIndex, true);
  }
  render() {
    const {
      account,
      accounts,
      actions,
      authorization,
      blockExplorers,
      data,
      pubkey,
      ledger,
      settings,
      system,
      t,
      transaction,
      wallets
    } = this.props;
    const {
      displayingKey,
      path,
      publicKey
    } = ledger;
    const {
      awaitingUpdate
    } = this.state;
    const accountData = accounts[account];
    const newkey = (publicKey) ? publicKey.wif : '';
    const wallet = find(wallets, {
      account,
      authorization,
    });
    let auth = {};
    const auths = new EOSAccount(accountData).getKeysForAuthorization(authorization);
    if (auths.length) {
      [auth] = auths;
    }
    let isConverting = false;
    let keysMatch = (newkey === auth.pubkey);
    if (wallet && wallet.convertParameters) {
      isConverting = true;
      keysMatch = (wallet.convertParameters.publicKey.wif === auth.pubkey);
    }
    const loading = (system.LEDGER_GET_PUBLIC_KEY === 'PENDING');
    const isValid = true;
    return (
      <Modal
        centered={false}
        closeIcon
        closeOnDimmerClick={false}
        onClose={this.onClose}
        open
        size
      >
        <Modal.Header>
          <Header
            content={t('ledger_account_convert_ledger_header')}
            icon="usb"
            subheader={t('ledger_account_convert_ledger_subheader')}
          />
        </Modal.Header>
        {(displayingKey)
          ? (
            <Modal.Content>
              <React.Fragment>
                <Header
                  content={t('ledger_account_convert_ledger_verify_header')}
                  subheader={t('ledger_account_convert_ledger_verify_subheader')}
                  textAlign="center"
                />
                <Segment
                  basic
                  size="large"
                  textAlign="center"
                >
                  {newkey}
                </Segment>
              </React.Fragment>
            </Modal.Content>
          )
          : false
        }
        {(loading)
          ? (
            <Modal.Content>
              <Header
                icon
                textAlign="center"
              >
                <Icon loading name="circle notched" />
                {t('ledger_account_convert_ledger_loading_key_header')}
                <Header.Subheader>
                  {t('ledger_account_convert_ledger_loading_key_subheader')}
                </Header.Subheader>
              </Header>
            </Modal.Content>
          )
          : false
        }
        {(!displayingKey && !newkey && !loading)
          ? (
            <React.Fragment>
              <Form
                as={Modal.Content}
              >
                <Message
                  content={t('ledger_account_convert_ledger_index_content')}
                  icon="help"
                  info
                  header={t('ledger_account_convert_ledger_index_header')}
                />
                <Form.Input
                  autoFocus
                  placeholder="0"
                  label={t('ledger_account_convert_ledger_index_label')}
                  name="index"
                  onChange={this.onChange}
                  value={this.state.index}
                />
              </Form>
              <Modal.Actions>
                <Form.Button
                  content={t('ledger_account_convert_ledger_key_button')}
                  onClick={this.loadKey}
                  primary
                />
              </Modal.Actions>
            </React.Fragment>
          )
          : false
        }
        {(!displayingKey && newkey && isConverting)
          ? (
            <React.Fragment>
              <Modal.Content>
                <Header
                  icon
                  textAlign="center"
                >
                  {(keysMatch)
                    ? (
                      <React.Fragment>
                        <Icon color="green" name="checkmark" />
                        {t('ledger_account_convert_ledger_converting_ready_header')}
                        <Header.Subheader>
                          {t('ledger_account_convert_ledger_converting_ready_subheader')}
                        </Header.Subheader>
                      </React.Fragment>
                    )
                    : (
                      <React.Fragment>
                        <Icon loading name="circle notched" />
                        {t('ledger_account_convert_ledger_converting_waiting_header')}
                        <Header.Subheader>
                          {t('ledger_account_convert_ledger_converting_waiting_subheader')}
                        </Header.Subheader>
                      </React.Fragment>
                    )
                  }
                </Header>
                <Table
                  definition
                  verticalAlign="middle"
                >
                  <Table.Header>
                    <Table.Row>
                      <Table.HeaderCell>Field</Table.HeaderCell>
                      <Table.HeaderCell>Value</Table.HeaderCell>
                      <Table.HeaderCell></Table.HeaderCell>
                    </Table.Row>
                  </Table.Header>
                  <Table.Body>
                    <Table.Row>
                      <Table.Cell>
                        {t('account_authority')}
                      </Table.Cell>
                      <Table.Cell>
                        {account}@{authorization}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_current')}
                      </Table.Cell>
                      <Table.Cell>
                        {auth.pubkey}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                        {(!keysMatch)
                          ? (
                            <Header
                              size="small"
                            >
                              <Header.Subheader>
                                {t('ledger_account_convert_ledger_key_updated')}
                              </Header.Subheader>
                              <TimeAgo date={`${accountData.head_block_time}z`} />
                            </Header>
                          )
                          : false
                        }
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_selected')}
                      </Table.Cell>
                      <Table.Cell>
                        {newkey}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_path')}
                      </Table.Cell>
                      <Table.Cell>
                        {path}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                      </Table.Cell>
                    </Table.Row>
                  </Table.Body>
                </Table>
              </Modal.Content>
              <Modal.Actions>
                {(!keysMatch)
                  ? (
                    <React.Fragment>
                      <Button
                        content={t('ledger_account_convert_ledger_converting_force_update')}
                        floated="left"
                        onClick={this.updateAccount}
                      />
                      {(wallet.mode !== 'hot')
                        ? (
                          <Button
                            content={t('ledger_account_convert_ledger_converting_reset')}
                            floated="left"
                            onClick={this.stopPrepare}
                          />
                        )
                        : false
                      }
                    </React.Fragment>
                  )
                  : false
                }
                {(isValid && !keysMatch && ['FAILURE', 'PENDING', 'SUCCESS'].includes(system.UPDATEAUTH))
                  ? (
                    <ToolsModalPermissionAuthSet
                      account={account}
                      actions={actions}
                      authorization={authorization}
                      auth={data}
                      blockExplorers={blockExplorers}
                      newkey={newkey}
                      onFormSubmit={this.onFormSubmit}
                      open
                      path={path}
                      pubkey={pubkey}
                      settings={settings}
                      system={system}
                    />
                  )
                  : false
                }
                {(isValid && !keysMatch && settings.walletMode === 'watch')
                  ? (
                    <WalletPanelButtonBroadcast
                      actions={actions}
                      blockExplorers={blockExplorers}
                      button={{
                        color: 'purple',
                        content: t('wallet:wallet_panel_wallet_broadcast'),
                        icon: 'wifi'
                      }}
                      settings={settings}
                      system={system}
                      transaction={transaction}
                    />
                  )
                  : false
                }
                {(keysMatch)
                  ? (
                    <Button
                      content={t('ledger_account_convert_ledger_converting_complete')}
                      onClick={this.completePrepare}
                      primary
                    />
                  )
                  : false
                }
              </Modal.Actions>
            </React.Fragment>
          )
          : false
        }
        {(!displayingKey && newkey && !isConverting)
          ? (
            <React.Fragment>
              <Modal.Content>
                <Message
                  content={t('ledger_account_convert_ledger_confirm_content')}
                  icon="help"
                  info
                  header={t('ledger_account_convert_ledger_confirm_header')}
                />
                <Table
                  definition
                  verticalAlign="middle"
                >
                  <Table.Header>
                    <Table.Row>
                      <Table.HeaderCell>Field</Table.HeaderCell>
                      <Table.HeaderCell>Value</Table.HeaderCell>
                      <Table.HeaderCell></Table.HeaderCell>
                    </Table.Row>
                  </Table.Header>
                  <Table.Body>
                    <Table.Row>
                      <Table.Cell>
                        {t('account_authority')}
                      </Table.Cell>
                      <Table.Cell>
                        {account}@{authorization}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_current')}
                      </Table.Cell>
                      <Table.Cell>
                        {auth.pubkey}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_selected')}
                      </Table.Cell>
                      <Table.Cell>
                        {newkey}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                        <Button
                          basic
                          content={t('ledger_account_convert_ledger_verify_key_button')}
                          icon="usb"
                          onClick={this.verify}
                          primary
                        />
                      </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                      <Table.Cell>
                        {t('ledger_account_convert_ledger_key_path')}
                      </Table.Cell>
                      <Table.Cell>
                        {path}
                      </Table.Cell>
                      <Table.Cell textAlign="right">
                        <Button
                          basic
                          content={t('ledger_account_convert_ledger_change_path_button')}
                          icon="pencil"
                          onClick={this.reset}
                          primary
                        />
                      </Table.Cell>
                    </Table.Row>
                  </Table.Body>
                </Table>
              </Modal.Content>
              <Modal.Actions>
                {(isValid && !keysMatch)
                  ? (
                    <ToolsModalPermissionAuthSet
                      account={account}
                      actions={actions}
                      authorization={authorization}
                      auth={data}
                      blockExplorers={blockExplorers}
                      button={{
                        color: 'purple',
                        content: t('ledger_account_convert_ledger_complete_button'),
                        fluid: false,
                        icon: 'lock',
                        size: 'small'
                      }}
                      newkey={newkey}
                      onFormSubmit={this.onFormSubmit}
                      path={path}
                      pubkey={pubkey}
                      settings={settings}
                      system={system}
                    />
                  )
                  : false
                }
                {(isValid && keysMatch)
                  ? (
                    <Button
                      content={t('ledger_account_convert_ledger_convert_wallet')}
                      onClick={this.onSubmit}
                      primary
                    />
                  )
                  : false
                }
              </Modal.Actions>
            </React.Fragment>
          )
          : false
        }
      </Modal>
    );
  }
}

function mapStateToProps(state) {
  return {
    accounts: state.accounts,
    blockExplorers: state.blockexplorers,
    ledger: state.ledger,
    settings: state.settings,
    status: HardwareLedgerActions.ledgerGetStatus(state.ledger),
    system: state.system,
    transaction: state.transaction,
    wallet: state.wallet,
    wallets: state.wallets
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...AccountActions,
      ...HardwareLedgerActions,
      ...SettingsActions,
      ...SystemStateActions,
      ...TransactionActions,
      ...UpdateAuthActions,
      ...WalletActions,
      ...WalletsActions,
    }, dispatch)
  };
}

export default compose(
  translate('ledger'),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalAccountConvertLedger);
