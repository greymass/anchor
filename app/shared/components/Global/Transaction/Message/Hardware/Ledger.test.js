import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalTransactionMessageHardwareLedger } from './Ledger';

const mockProps = {
  t: text => text,
};

describe('GlobalTransactionMessageHardwareLedger', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalTransactionMessageHardwareLedger {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
