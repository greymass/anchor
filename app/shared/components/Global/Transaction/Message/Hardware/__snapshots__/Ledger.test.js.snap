// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalTransactionMessageHardwareLedger renders correctly 1`] = `
<Segment
  basic={true}
>
  <Header
    icon={true}
    size="large"
    textAlign="center"
  >
    <Icon
      as="i"
      loading={true}
      name="circle notched"
    />
    global_transaction_hardware_ledger_awaiting_device_header
    <HeaderSubheader
      content="global_transaction_hardware_ledger_awaiting_device_subheader"
    />
  </Header>
</Segment>
`;
